text/0000-adding-pyo3-like-attribute-macros.md

```diff
@@ -0,0 +1,75 @@
+- Feature Name: adding_pyo3_like_attribute_macros_for_classes_and_methods
+- Start Date: 2021-06-02
+- RFC PR: (leave this empty)
+- Neon Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+This RFC proposes adding [PyO3-like](https://pyo3.rs/v0.13.2/class.html) attribute-macros to define JavaScript classes as Rust structs. 
+
+# Motivation
+[motivation]: #motivation
+
+At the moment of migration to `napi`, there is no idiomatic way of writing JavaScript classes in pure Rust. This proposal adds the attribute macros to help users to write classes and methods easily. 
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+I will describe some potential use cases here:
+```rust
+#[neon::class]
+struct Person {
+    name: string,
+    age: u8,
+}
+#[neon::methods]
+impl Person {
+    fn greet(&self, cx: FunctionContext) -> JsResult<JsString> {
+        Ok(cx.string(format!("Hello, my name is {}", self.name)))
+    }
+
+    #[neon::constructor]
```

Comment:
The current `declare_types` macro has a two-part initialization with the second part being optional.

1. (Required). Constructor. This creates the Rust struct that gets wrapped into the JS class.
2. (Optional). Initialization. This has access to the _created_ class and can perform actions like assigning properties. This can be thought of as using `this` in a JS constructor.

I think this is valuable and it would be nice to maintain that functionality. I can think of a couple ways to implement it:

1. Another macro attribute like `#[neon::class_init]` for that phase
2. Need to call a function and return a class instead of the data type.

```rust
create_class(Person { name, age })
```